src/nvim/ex_cmds.c

```diff
@@ -1056,6 +1058,20 @@ void do_bang(int addr_count, exarg_T *eap, bool forceit, bool do_in, bool do_out
     STRCAT(newcmd, prevcmd);
     STRCAT(newcmd, p_shq);
     free_newcmd = true;
+  } else {
+    if (addr_count != 0) {
```

Comment:
Seems like this should go before the `if (*p_shq != NUL)` check (which then becomes `else if (*p_shq != NUL)`).